socs/agents/hwp_pid/drivers/pid_controller.py

```diff
@@ -45,18 +47,15 @@ def _establish_connection(ip, port, timeout=5):
 
         """
         conn = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        conn.settimeout(timeout)
         # unit tests might fail on first connection attempt
-        attempts = 3
-        for attempt in range(attempts):
+        while True:
```

Comment:
I don't love this being an infinite loop, since control is never passed back to the agent. Can we keep this to a finite number of attempts, and handle re-connection at the agent level?